src/base/materialized.ts

```diff
@@ -230,60 +236,203 @@ export class MaterializedChatMessage implements IMaterializedNode {
+					if (child instanceof MaterializedChatMessageTextChunk) {
+						result.push(child.text);
+					}
+				}
+				break;
+			}
 			if (
 				text.lineBreakBefore === LineBreakBefore.Always ||
 				(text.lineBreakBefore === LineBreakBefore.IfNotTextSibling && !isTextSibling)
 			) {
-				if (result.length && !result.endsWith('\n')) {
-					result += '\n';
+				let res = result[result.length - 1]
+				if (!(res instanceof MaterializedChatMesageImage) && result.length && !res.endsWith('\n')) {
+					result.push('\n');
 				}
 			}
 
-			result += text.text;
+			if (typeof result[result.length - 1] === 'string') {
+				result[result.length - 1] += text.text;
+			} else {
+				result.push(text.text);
+			}
 		}
 
-		return result.trim();
+		return result;
 	});
 
 	public toChatMessage(): ChatMessage {
+		const content = this.text
+			.filter(element => typeof element === 'string')
+			.join('').trim();
+
+		const images = this._text().filter(element => element instanceof MaterializedChatMesageImage) as MaterializedChatMesageImage[];
+
+		if (images.length > 0) {
+			return {
+				role: ChatRole.User,
+				content: [
+					{ type: 'text', text: content },
+					{
+						type: 'image_url',
+						image_url: { url: getEncodedBase64(images[0].imageUrl), detail: images[0].detail },
+					}],
+			};
+		}
+
 		if (this.role === ChatRole.System) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				...(this.name ? { name: this.name } : {}),
 			};
 		} else if (this.role === ChatRole.Assistant) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				...(this.toolCalls ? { tool_calls: this.toolCalls } : {}),
 				...(this.name ? { name: this.name } : {}),
 			};
 		} else if (this.role === ChatRole.User) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				...(this.name ? { name: this.name } : {}),
 			};
 		} else if (this.role === ChatRole.Tool) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				tool_call_id: this.toolCallId,
 			};
 		} else {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				name: this.name!,
 			};
 		}
 	}
 }
 
+export class MaterializedChatMesageImage implements IMaterializedNode {
+	constructor(
+		public readonly id: number,
+		public readonly role: ChatRole,
+		public readonly imageUrl: string,
+		public readonly detail: 'low' | 'high',
+		public readonly priority: number,
+		public readonly metadata: PromptMetadata[] = [],
+		public readonly lineBreakBefore: LineBreakBefore,
+		public readonly children: MaterializedNode[]
+	) {}
+	upperBoundTokenCount(tokenizer: ITokenizer): Promise<number> {
+		return this.upperBoundTokenCount(tokenizer);
+	}
+	tokenCount(tokenizer: ITokenizer): Promise<number> {
+		return this._tokenCount(tokenizer);
+	}
+
+	private readonly _tokenCount = once(async (tokenizer: ITokenizer) => {
+		return tokenizer.countMessageTokens(this.toChatMessage());
```

Comment:
I think counting this as a chat message is not the right thing to do. This is a _part_ of a message (at least based on the above logic and AI APIs) and counting it as its own entire message will overcount the number of tokens it would actually consume as one part of a message.

I don't think you need `toChatMessage` in this class at all.